subprojects/language-native/src/integTest/groovy/org/gradle/language/AbstractNativeParallelIntegrationTest.groovy

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.language
+
+import org.gradle.api.specs.Spec
+import org.gradle.integtests.fixtures.BuildOperationsFixture
+import org.gradle.internal.execution.ExecuteTaskBuildOperationType
+import org.gradle.internal.operations.trace.BuildOperationRecord
+import org.gradle.nativeplatform.fixtures.AbstractInstalledToolChainIntegrationSpec
+
+abstract class AbstractNativeParallelIntegrationTest extends AbstractInstalledToolChainIntegrationSpec {
+    BuildOperationsFixture buildOperations = new BuildOperationsFixture(executer, temporaryFolder)
+
+    def setup() {
+        executer.withArgument("--max-workers=4")
+    }
+
+    void assertTaskIsParallel(String taskName) {
+        def task = buildOperations.first(ExecuteTaskBuildOperationType, new Spec<BuildOperationRecord>() {
+            @Override
+            boolean isSatisfiedBy(BuildOperationRecord record) {
+                return record.displayName == "Task :${taskName}"
+            }
+        })
+
+        assert task != null
+        assert buildOperations.getOperationsConcurrentWith(ExecuteTaskBuildOperationType, task).size() > 0
+    }
+
+    def withTaskThatRunsParallelWith(String taskName) {
+        buildFile << """
+            import java.util.concurrent.CountDownLatch
+            import org.gradle.internal.work.WorkerLeaseService
+
+            def taskFinished = new CountDownLatch(1)
+            gradle.taskGraph.afterTask { task ->
+                if (task.name == "${taskName}") {
+                    taskFinished.countDown()
+                }
+            }
+            
+            task parallelTask {
+                dependsOn { tasks.${taskName}.taskDependencies }
+                doLast { 
+                    services.get(WorkerLeaseService).withoutProjectLock {
```

Comment:
Ok, per our conversation, I've implemented a somewhat dirty solution that registers a custom toolchain on the task that we are trying to test.  This custom toolchain delegates to the "real" toolchain except that it sets a decorated `BuildOperationExecutor` that allows us to hook into the moment right before and after the compiling/linking is run.  This allows us to accomplish 2 and 3 above by synchronizing `parallelTask` with the compiling/linking steps and ensure that it runs in parallel.  The custom toolchain isn't the prettiest solution, but it does seem to work.  If this seems like too much of a hack, we can pull it out and rely on a performance test as we discussed.

Item 1 above is currently accomplished by ordering `parallelTask` on the command line after the assemble task (i.e. taking advantage of the fact that this will put it later in the execution queue).  This ordering is admittedly weak (changes to the way we build the execution graph/queue could break this), but I don't see another reasonable way to accomplish item 1 with our current mechanisms.